pkg/collect/cluster_resources.go

```diff
@@ -156,190 +166,144 @@ func (c *CollectClusterResources) Collect(progressChan chan<- interface{}) (Coll
-	for k, v := range pvcs {
-		output.SaveResult(c.BundlePath, path.Join("cluster-resources/pvcs", k), bytes.NewBuffer(v))
-	}
-	output.SaveResult(c.BundlePath, "cluster-resources/pvcs-errors.json", marshalErrors(pvcsErrors))
-
-	//Roles
-	roles, rolesErrors := roles(ctx, client, namespaceNames)
-	for k, v := range roles {
-		output.SaveResult(c.BundlePath, path.Join("cluster-resources/roles", k), bytes.NewBuffer(v))
-	}
-	output.SaveResult(c.BundlePath, "cluster-resources/roles-errors.json", marshalErrors(rolesErrors))
-
-	//Role Bindings
-	roleBindings, roleBindingsErrors := roleBindings(ctx, client, namespaceNames)
-	for k, v := range roleBindings {
-		output.SaveResult(c.BundlePath, path.Join("cluster-resources/rolebindings", k), bytes.NewBuffer(v))
-	}
-	output.SaveResult(c.BundlePath, "cluster-resources/rolebindings-errors.json", marshalErrors(roleBindingsErrors))
+	clusterResourceCollectors = append(clusterResourceCollectors, func() {
+		pvs, pvsErrors := pvs(ctx, client)
+		output.SaveResult(c.BundlePath, "cluster-resources/pvs.json", bytes.NewBuffer(pvs))
+		output.SaveResult(c.BundlePath, "cluster-resources/pvs-errors.json", marshalErrors(pvsErrors))
+	})
 
 	//Cluster Roles
-	clusterRoles, clusterRolesErrors := clusterRoles(ctx, client)
-	output.SaveResult(c.BundlePath, "cluster-resources/clusterroles.json", bytes.NewBuffer(clusterRoles))
-	output.SaveResult(c.BundlePath, "cluster-resources/clusterroles-errors.json", marshalErrors(clusterRolesErrors))
+	clusterResourceCollectors = append(clusterResourceCollectors, func() {
+		clusterRoles, clusterRolesErrors := clusterRoles(ctx, client)
+		output.SaveResult(c.BundlePath, "cluster-resources/clusterroles.json", bytes.NewBuffer(clusterRoles))
+		output.SaveResult(c.BundlePath, "cluster-resources/clusterroles-errors.json", marshalErrors(clusterRolesErrors))
+	})
 
 	//Cluster Role Bindings
-	clusterRoleBindings, clusterRoleBindingsErrors := clusterRoleBindings(ctx, client)
-	output.SaveResult(c.BundlePath, "cluster-resources/clusterRoleBindings.json", bytes.NewBuffer(clusterRoleBindings))
-	output.SaveResult(c.BundlePath, "cluster-resources/clusterRoleBindings-errors.json", marshalErrors(clusterRoleBindingsErrors))
+	clusterResourceCollectors = append(clusterResourceCollectors, func() {
+		clusterRoleBindings, clusterRoleBindingsErrors := clusterRoleBindings(ctx, client)
+		output.SaveResult(c.BundlePath, "cluster-resources/clusterRoleBindings.json", bytes.NewBuffer(clusterRoleBindings))
+		output.SaveResult(c.BundlePath, "cluster-resources/clusterRoleBindings-errors.json", marshalErrors(clusterRoleBindingsErrors))
+	})
 
+	// run the collectors
+
+	maxNbConcurrentGoroutines := 5
+	concurrentGoroutines := make(chan struct{}, maxNbConcurrentGoroutines)
+
+	for _, thisClusterResourceCollector := range clusterResourceCollectors {
+		wg.Add(1)
+		go func(crCollector getClusterResource) {
+			defer wg.Done()
+			concurrentGoroutines <- struct{}{}
```

Comment:
this pattern kicks off all the goroutines at once, but blocks the actual work till the channel has a free slot